### PR TITLE
prune correctly when series is both leaf and branch

### LINKED ISSUE #797: statMetricsActive is incorrect because we prune recursively.
see https://github.com/grafana/metrictank/pull/787 more specifically https://github.com/grafana/metrictank/pull/787#pullrequestreview-84874181

copying here because I don't trust GH enough to retain the comments long after the fact.

in our implementation, a path can be both a leaf and have children.
e.g. `foo.bar` can have data/metricdef (making it a leaf), while `foo.bar.baz` also exists (making foo.bar a branch)
In graphite this doesn't work, but some of our customers wanted this behavior, and it seemed fairly easy to support, so that's why we do.

the `m.delete` call is recursive: it deletes the path, and any sub-paths ([I think i actually just discovered a related bug here](https://github.com/grafana/metrictank/issues/714#issuecomment-353354166)), so what we should do is track the actual amount of paths that got deleted by the delete call, or I think better: the pruning process shouldn't recursively delete, but rather only delete leaves (and clean up stale branch nodes that no longer have children nor defs)

### PR DESCRIPTION
**Pull request by woodsaj**
-  this issue was discovered in discussions of issue #714
-  this also fixes #797

### CODE REVIEW

File: idx/memory/memory.go [hunk 1]
```diff
@@ -1136,8 +1136,24 @@ func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]idx.Archive, error) {
...
+					log.Debug("memory-idx: deleting %s from index", id)
+					defs = append(defs, *m.DefById[id])
+					delete(m.DefById, id)
+				}
+				n.Defs = []string{}
+				if tagSupport {
+					for _, def := range defs {
+						m.deindexTags(&(def.MetricDefinition))
+					}
+				}
```
Comment (DanCech) [inline idx/memory/memory.go]: Would it make sense to add a `deleteChildren` parameter to `m.delete` (and have `m.Delete` pass `true`)? It seems like there would be only a small amount of logic needed to have it handle this and would cut down on repeated code.

**Comment by woodsaj** [comment]
per @DanCech 's suggestion, i refactored to move the logic into the MemoryIdx.delete() method.

File: idx/memory/memory.go [hunk 2]
```diff
@@ -1136,8 +1142,8 @@ func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]idx.Archive, error) {
 			}
 
 			log.Debug("memory-idx: series %s for orgId:%d is stale. pruning it.", n.Path, org)
-			defs := m.delete(org, n, true)
-			statMetricsActive.Dec()
+			defs := m.delete(org, n, true, false)
+			statMetricsActive.Add(-1 * len(defs))
```
Comment (Dieterbe) [inline idx/memory/memory.go]: the gauge has a `Dec` method

File: idx/memory/memory.go [hunk 3]
```diff
@@ -980,7 +980,7 @@ func (m *MemoryIdx) Delete(orgId int, pattern string) ([]idx.Archive, error) {
 	}
 
 	for _, f := range found {
-		deleted := m.delete(orgId, f, true)
+		deleted := m.delete(orgId, f, true, true)
```
Comment (Dieterbe) [inline idx/memory/memory.go]: is this the behavior we want? let's say you have data under these 2 paths:
```
foo.bar
foo.bar.baz
```
afaik a render query for `foo.*` would only return the first one, so a delete should probably also only delete the first one.

IOW is there even a use case at all for recursive deletes?

Comment (DanCech) [reply idx/memory/memory.go]: if you want a recursive delete then `foo.**` should match them all, so I'd tend to agree that the user should be required to be explicit.

Comment (Dieterbe) [reply idx/memory/memory.go]: > if you want a recursive delete then foo.** should match them all

is there any existing implementation/use of this format? it seems sensible.

Comment (DanCech) [reply idx/memory/memory.go]: It's not documented in https://graphite.readthedocs.io/en/latest/render_api.html#target (I'll have to fix that) but is supported by the standard whisper/rrd finder in graphite-web https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/finders/standard.py#L97

**Pull request merged**